Integrate Birco V2

AdnanElAssadi56 opened the pull request:
- Added a new task class BIRCOReRankingTask, extending the AbsTask base class.
- Implemented the necessary methods for task initialization, evaluation, and metadata retrieval.
- Integrated BIRCO task logic to support reranking evaluations.

Closes https://github.com/embeddings-benchmark/mteb/issues/818

AdnanElAssadi56 commented:
@orionw Can you please check this?

Samoed commented:
I've changed yor base branch to v2, because in your history was a lot of PRs from there.

AdnanElAssadi56 commented:
> I've changed yor base branch to v2, because in your history was a lot of PRs from there.

Thanks, I was just about to do that. Feel free to also give it a look. It is quite different from the previous version because v2 already has many of the implementations.

Samoed commented:
This PR and https://github.com/embeddings-benchmark/mteb/pull/1947 adding same tasks? If so maybe old PR needs to be closed?

Samoed commented on mteb/tasks/Reranking/eng/BIRCO_Reranking.py:
```diff
@@ -0,0 +1,211 @@
+from ....abstasks.AbsTaskReranking import AbsTaskReranking
+from mteb.abstasks.TaskMetadata import TaskMetadata
+
+class BIRCORerankingBase(AbsTaskReranking):
+    """
+    Base class for all BIRCO reranking tasks.
+    
+    Expects the Hugging Face dataset to have records with keys:
+      - "query_id": str
+      - "query_text": str
+      - "corpus": list of dicts (each with "corpus_id" and "corpus_text")
+      - "relevance": dict mapping corpus_id to an integer relevance score
+    Also uses an instruction stored in metadata (metadata.instruction).
+    """
+    def dataset_transform(self, hf_dataset):
+        queries = {}
+        corpus = {}
+        qrels = {}
+        instructions = {}
+        
+        # For each record, build the standard dictionaries.
+        for record in hf_dataset:
+            qid = record["query_id"]
+            queries[qid] = record["query_text"]
+            qrels[qid] = record["relevance"]
+            # Use the instruction defined in metadata (if provided); otherwise, use an empty string.
+            instructions[qid] = getattr(self.metadata, "instruction", "")
+            for doc in record["corpus"]:
+                cid = doc["corpus_id"]
+                if cid not in corpus:
+                    corpus[cid] = doc["corpus_text"]
+                    
+        return {"queries": queries, "corpus": corpus, "qrels": qrels, "instructions": instructions}
```
Can you change to function, because it is a bit hard to deal with classes that not tasks

Samoed replied:
Retrieval format have different data structure and I think this won't run. Also can you reupload data in standard format? You can do this by:
```python
import mteb

task = mteb.get_task("task_name")
task.push_dataset_to_hub("mteb/{dataset_name}")
```

Samoed replied:
Still can you remove this class?

Samoed replied:
You should remove `BIRCORerankingBase`

Samoed commented on mteb/tasks/Reranking/eng/BIRCO_Reranking.py:
```diff
@@ -0,0 +1,211 @@
+        eval_splits=["test"],
+        eval_langs=["eng-Latn"],
+        main_score="ndcg",
+        dataset={"path": "mteb/BIRCO-ArguAna-Test", "revision": "latest"},
```
Specify exact revision of the dataset

Samoed commented on mteb/tasks/Reranking/eng/BIRCO_Reranking.py:
```diff
@@ -0,0 +1,211 @@
+        eval_splits=["test"],
+        eval_langs=["eng-Latn"],
+        main_score="ndcg",
+        dataset={"path": "mteb/BIRCO-DorisMae-Test", "revision": "latest"},
```
Please, specify exact revision here as commit revision

Samoed commented on mteb/tasks/Reranking/eng/BIRCO_Reranking.py:
```diff
@@ -0,0 +1,211 @@
+        modalities=["text"],
+        eval_splits=["test"],
+        eval_langs=["eng-Latn"],
+        main_score="ndcg",
```
We don't have `ndcg` metric. I think this should be as
```suggestion
        main_score="ndcg_at_10",
```

Samoed commented on mteb/tasks/Reranking/eng/BIRCO_Reranking.py:
```diff
@@ -0,0 +1,211 @@
+            queries[qid] = record["query_text"]
+            qrels[qid] = record["relevance"]
+            # Use the instruction defined in metadata (if provided); otherwise, use an empty string.
+            instructions[qid] = getattr(self.metadata, "instruction", "")
```
If your task if instruction retrieval, then add instructions to dataset directly

Samoed reviewed (commented):
Can you run tasks to make sure that score of tasks is matching with author's implementation?

Samoed commented on .vscode/settings.json:
Please, don't delete `.vscode` folder

Samoed commented on mteb/tasks/Reranking/eng/BIRCOArguAnaReranking.py:
```diff
@@ -0,0 +1,67 @@
+    instruction = "Given a one-paragraph argument, retrieve the passage that contains the counter-argument which directly refutes the query's stance."
+
+    def dataset_transform(self, hf_dataset):
+        return birco_transform(hf_dataset, self.instruction)
```
I think that you should add 
```suggestion
    transformed_birco = birco_transform(hf_dataset, self.instruction)
    self.query = transformed_birco["query"]
    self.corpus = transformed_birco["corpus"]
    self.qrels = transformed_birco["qrels"]
```

Samoed commented on mteb/tasks/Reranking/eng/BIRCOArguAnaReranking.py:
```diff
@@ -0,0 +1,67 @@
+  howpublished={\\url{https://github.com/BIRCO-benchmark/BIRCO}}
+}""",
+    )
+    instruction = "Given a one-paragraph argument, retrieve the passage that contains the counter-argument which directly refutes the query's stance."
```
I'm not sure that `instruction` should be used like that. I think that should be transformed to prompt

Samoed commented on mteb/tasks/Reranking/eng/BIRCODorisMaeReranking.py:
```diff
@@ -0,0 +1,70 @@
+        annotations_creators="expert-annotated",  # Valid annotator type
+        dialect=[],
+        sample_creation="found",  # Valid sample creation method
+        bibtex_citation="""@misc{BIRCO,
+  title={BIRCO: A Benchmark of Information Retrieval Tasks with Complex Objectives},
+  author={Wang et al.},
+  year={2024},
+  howpublished={\\url{https://github.com/BIRCO-benchmark/BIRCO}}
+}""",
```
I think citation should be changed to this every where
```suggestion
        bibtex_citation="""@misc{wang2024bircobenchmarkinformationretrieval,
      title={BIRCO: A Benchmark of Information Retrieval Tasks with Complex Objectives}, 
      author={Xiaoyue Wang and Jianyou Wang and Weili Cao and Kaicheng Wang and Ramamohan Paturi and Leon Bergen},
      year={2024},
      eprint={2402.14151},
      archivePrefix={arXiv},
      primaryClass={cs.IR},
      url={https://arxiv.org/abs/2402.14151}, 
}""",

```

AdnanElAssadi56 commented:
Hi @Samoed 

I've refactored the datasets. I preprocessed them externally, and now they look like the retrieval/reranking datasets on HF. Hence, I've deleted all the transforming/loading. I also fixed formatting like citations and other points you mentioned.

They should work now.

For some reason, I cannot test them locally. The first time I ran the tests, they failed because they expected corpus, queries, and qrels as configs instead of default, so I did that. 

When I try to run tests now, it says it expects default as a config, so I am not sure.

Samoed commented:
I think issue with eval splits. In TaskMetadata you've annotated that they will use "test" split, but on huggingface you've uploaded train

Samoed reviewed (commented):
Can you please get back the `.vscode` folder and upload the dev and test splits for the datasets, as it is not currently clear what "train" is, as it was not in the original dataset?

AdnanElAssadi56 commented:
I added the vscode folder. 

There is only a test split for the dataset as far as I am aware because we are just evaluating. 
It now shows the split as test on huggingface.

Can you please run it from your side and tell me if it works?

Samoed commented:
Yes, I'll try to run it. Can you resolve merge conflic?

Samoed commented:
Ah, yes. You should also calculate `task.calculate_metadata_metrics()` for each of the task

AdnanElAssadi56 commented:
Do I add them in the classes for each task? And under which function? 
Why isn't the calculation done in the abstract class, or is it just a one time thing?

Samoed commented:
You should call this function for each of the tasks. This function will produce json in `descriptive_stats` folder with info about dataset.

AdnanElAssadi56 commented:
I am getting this error when I try to calculate metrics:
ValueError: BuilderConfig 'default' not found. Available: ['corpus', 'queries', 'qrels']

Isn't it expected to have corpus, queries, qrels as configs?

AdnanElAssadi56 commented:
I tried changing qrels to default because apparently some datasets have it this way, but it is giving me this:
ValueError: Couldn't find cache for mteb/BIRCO-DorisMae-Test for config 'default'
Available configs in the cache: ['corpus', 'queries']

Can you please tell me what structure is the correct one?

Samoed commented:
I can download `BIRCO-DorisMae` without issues with `010fedd0de72e8b77388ea5b1d563d638d5900e5` revision

Samoed commented on mteb/tasks/Reranking/eng/BIRCODorisMaeReranking.py:
```diff
@@ -0,0 +1,44 @@
+        date=("2024-01-01", "2024-12-31"),
+        domains=["Academic"],  # Valid combination
+        task_subtypes=["Scientific Reranking"],  # MTEB-approved subtype
+        license="https://creativecommons.org/licenses/by/4.0/",  # Full URL
```
```suggestion
        license="cc-by-4.0",
```

AdnanElAssadi56 commented:
> I can download `BIRCO-DorisMae` without issues with `010fedd0de72e8b77388ea5b1d563d638d5900e5` revision

I still haven't commited that revision. I've only made the change on HF. This revision has default as a config instead of qrels. Is this how it's supposed to be?

Samoed commented:
Yes, that dataset has right split format

AdnanElAssadi56 commented:
Fixed License and now all datasets have default, queries, and corpus as configs.

AdnanElAssadi56 commented:
Hi @Samoed 

Another dataset seems to be causing the tests to fail "Namaa." Can you please confirm?

Samoed commented:
Yes, but that a bit weird, because I can find it on HF

Samoed reviewed (approved):
Great work!
